aten/src/ATen/VmapTransforms.h

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include <ATen/BatchedTensorImpl.h>
+
+namespace at {
+
+// This file contains abstractions used for transforming *logical* vmap arguments
+// into *physical* arguments. (Keep reading for definitions of these terms).
+
+// NOTE: [Logical vs physical args]
+// Consider the following vmap.
+//   vmap(vmap(func, in_dims=(2,)), in_dims=(0,))(torch.ones(2, 3, 4))
+// This would produce a BatchedTensor wrapping a Tensor of size [2, 3, 4],
+// with batch dims 0 and 2:
+//   BatchedTensor(ones(2, 3, 4), bdims=[(lvl=1,dim=0),(lvl=2,dim=2)])
+//
+// We say the *logical* view of the tensor has size [3] -- tensors inside
+// `func` appear to have size [3].
+// However, the *physical* underlying tensor (the one passed to vmap) has size
+// [2, 3, 4].
+//
+// This notion of logical vs physical also extends to non-tensor arguments.
+// Consider the previous tensor; let's assume the user called
+// `torch.sum(tensor, dim=0)` inside of `func`. Then the logical
+// dimension they are reducing over is dim 0 but the physical dim is dim 1
+// (the first non-batch dimension)
+
+// Forward declared; see NOTE: [What is a PhysicalView?]
+struct PhysicalView;
+
+// NOTE: [What is an VmapTransform?]
+// An *VmapTransform* converts logical views of tensors to physical views.
+//
+// Batching rules use VmapTransforms to convert logical arguments to
+// physical arguments, then call one or more at:: operator that handles the
+// physical arguments, and then converts the physical result back to a logical
+// argument.
+
+// VmapTransform for operators that take tensors with multiple batch dims.
+// Given one or more logical views on Tensors, `logicalToPhysical` 
+// permutes all of the batch dims to the front of the tensor, aligns
+// and expands the batch dims to match each other (according to their `level`),
+// and returns a PhysicalView on the tensor(s).
+struct TORCH_API MultiBatchVmapTransform {
+  static PhysicalView logicalToPhysical(const Tensor& logical_tensor);
+  static std::vector<PhysicalView> logicalToPhysical(TensorList logical_tensors);
+};
+
+// NOTE: [What is a PhysicalView?]
+// PhysicalView represents a physical view on a Tensor.
+//
+// One can use it to further convert logical dimension indices, logical shapes,
+// and more to their physical variants, or convert a new (physical) tensor into
+// a logical BatchedTensor. (TODO(rzou): some of these are not yet implemented).
+//
+// PhysicalView stores a physical tensor with all of its batch dimensions at
+// the front and some levels that correspond to said batch dimensions.
+//
+// The levels contain the information necessary to perform the logical<->physical
+// mapping. For example, given:
```

Comment:
This is a really indirect way to describe what "levels" is. Don't say what the information is used to do: tell me what the information means. "The levels bitset specifies which vmap levels correspond to the batch dimensions at the front of the tensor. In particular, the number of set bits corresponds to the number of batch dimensions on tensor, and the rightmost bit of levels specifies the minimum number of nested vmaps we are in at this point of time."

Also, this particular representation of PhysicalView doesn't seem to generalize to other vmap transforms. For example, if you have a tensor that accepts only one batch dimension, you're going to need to store something else in this case, right?

Reply:
> Also, this particular representation of PhysicalView doesn't seem to generalize to other vmap transforms. For example, if you have a tensor that accepts only one batch dimension, you're going to need to store something else in this case, right?

Yeah, you'd have to store levels and the batch sizes. I'm not sure if I should reuse the PhysicalView class when that happens (and add an optional<batch_sizes> field), or just make a different class for it and call it "FlattenedBatchPhysicalView"

Reply:
How about: PhysicalView ties the packed data Tensor underlying a BatchedTensor with a bitset annotates the dimensions of the packed data Tensor as either user visible (if the corresponding bit is set to 1) or implementation-only visible (if the corresponding bit is set to 0).